### DISCUSSION TITLE
Usage with Kafka + Apicurio Schema Registry

### DISCUSSION THREAD

**Discussion by CROprogrammer** [microcks/microcks#845] [Q&A] [answered]
Hi there,

I have couple of questions about usage of Microcks with Kafka and Apicurio Schema Registry.
How do I set to which Kafka topic will messages write to? I want to specify my own topic, not that microcks generates its name from AsyncAPI schema name.
Can I retrieve schemas from my schema registry to Microcks so that I don't have to import AsyncAPI schemas by myself?
How does Microcks use schema registry? Does it only use it to register schemas, does it use registry to validate schemas, etc.?
I don't want to send random Kafka messages every X seconds, I just want to send one with specified values, and I don't want to upload new AsyncAPI schema everytime I do it. Is that possible?
Can I send Kafka messages with Avro schema via Microcks integration with Postman? Can I send Kafka messages via Postman?

**Comment by CROprogrammer** [comment]
One more question, 
does Microcks create Kafka topic when the first message arrives?

**Reply by lbroudoux** [reply, marked as answer]
Hey @CROprogrammer

That's a lot of questions ... Will try not to forget one 😉 

> How do I set to which Kafka topic will messages write to? I want to specify my own topic, not that microcks generates its name from AsyncAPI schema name. Does Microcks create Kafka topic when the first message arrives?

You cannot set and specify the Kafka topic Microcks creating/using for sending mock messages at the moment. Topic is crated by Microcks at first mock message publication time. The name is based on "API name + API version + channel name" and is determined in a way to avoid collision with other APIs or versions. Thus developer can still receive events from the 1.0 version when 1.1 is available and so on... We'd like to make this naming strategy configurable in the future but didn't start working on that yet.

> Can I retrieve schemas from my schema registry to Microcks so that I don't have to import AsyncAPI schemas by myself?

If you store the whole AsyncAPI specification in your registry and this specification can be retrieved through a single HTTP call, this should be possible to [configure an importer](https://microcks.io/documentation/using/importers/#creating-a-new-scheduled-import) for that.

> How does Microcks use schema registry? Does it only use it to register schemas, does it use registry to validate schemas, etc.?

Schema registry can be used in two different use cases:
1. For mocking: Microcks can push for example Avro schema (referenced or contained into AsyncAPI spec) to the registry at the time it's serializing a mock message and sending it to the topic it's managing
2. For testing: Microcks can connect to a remote schema registry associated with the Kafka broker is connected to in order to received and validate messages. That way, the Avro schema found in registry is used to deserialize and validate the messages that have been read.

We tried to illustrate both use cases in this [guide](https://microcks.io/documentation/guides/avro-messaging/).

> I don't want to send random Kafka messages every X seconds, I just want to send one with specified values, 

Not at the moment. The async dispatcher is quite dumb and send every example messages at each iteration. However, we started discussing about more more smart dispatcher on even "on demand" dispatcher with the community users (see https://microcksio.zulipchat.com/#narrow/stream/256143-users/topic/AsyncAPI.20Examples.20scheduling/near/362663790)

> and I don't want to upload new AsyncAPI schema everytime I do it. Is that possible?

If you created an importer, the scheduled refresh should prevent you from that. Also, you can decide to explicitly push your AsyncAPI spec update on each and every commit with a trigger that uses the [microcks-cli](https://microcks.io/documentation/automating/cli/#import-command) or other CI/CD extensions.

> Can I send Kafka messages with Avro schema via Microcks integration with Postman? Can I send Kafka messages via Postman?

Nothing that I am aware of regarding the Kafka support in Postman.

I hope it helps you to have a quick overview of Microcks capabilities. And as we're an open source project, please feel free to discuss and contribute ideas on your use cases!

Cheers,
